consensus/istanbul/backend/engine.go

```diff
@@ -645,8 +645,6 @@ func (sb *Backend) SetCallBacks(hasBadBlock func(common.Hash) bool,
 	processBlock func(*types.Block, *state.StateDB) (types.Receipts, []*types.Log, uint64, error),
 	validateState func(*types.Block, *state.StateDB, types.Receipts, uint64) error,
 	onNewConsensusBlock func(block *types.Block, receipts []*types.Receipt, logs []*types.Log, state *state.StateDB)) error {
-	sb.coreMu.RLock()
-	defer sb.coreMu.RUnlock()
```

Comment:
I think this usage of `coreMu` needs to remain, as it ensures that there is no data race between the values set in `SetCallbacks` and when they are read in `StartValidating`.

Currently `SetCallbacks`  & `StartValidating` are called by the same thread so it would be safe, but there is nothing that enforces that.

I think a better solution would be to pass the callbacks to StartValidating, that avoids any possibility of a data race over those fields.